Draft docs platform repositioning (#4915)
neondatabase/website: content/docs/get-started/full-backend-quickstart.md

</TwoColumnLayout.Block>
<TwoColumnLayout.Block>

<Tabs labels={["Console", "API"]}>

<TabItem>

In the project sidebar, go to **Auth** and click **Enable Auth**. On the **Configuration** tab, copy your **Auth URL**.

![Neon Auth Base URL](/docs/auth/neon-auth-base-url.png)

</TabItem>

<TabItem>

You'll need your project ID and default branch ID. If you used the API in step 1, the response contains both. Otherwise list projects to find them:

```bash filename="Terminal"
curl https://console.neon.tech/api/v2/projects \
  -H "Authorization: Bearer $NEON_API_KEY"
```

Enable Auth on the default branch:

```bash filename="Terminal"
curl -X POST \
  "https://console.neon.tech/api/v2/projects/$PROJECT_ID/branches/$BRANCH_ID/auth" \
  -H "Authorization: Bearer $NEON_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{"auth_provider": "better_auth"}'
```

The Auth URL is in the response under `jwks_url` (strip the `/.well-known/jwks.json` suffix).

</TabItem>

</Tabs>

</TwoColumnLayout.Block>
</TwoColumnLayout.Step>

<TwoColumnLayout.Step title="Scaffold a Next.js app">
<TwoColumnLayout.Block>

> Create a new Next.js project with TypeScript, Tailwind, and the App Router. The `--yes` flag accepts the remaining defaults (Turbopack, import alias) without prompting.

Reviewer: ```suggestion
Create a new Next.js project with TypeScript, Tailwind CSS, and the App Router. The `--yes` flag accepts the remaining defaults without prompting.
```

DATABASE_URL=postgresql://...
NEON_AUTH_BASE_URL=https://ep-xxx.neonauth.c-7.us-east-1.aws.neon.tech/neondb/auth
NEON_AUTH_COOKIE_SECRET=replace-with-32-char-random-secret
```

</TwoColumnLayout.Block>
</TwoColumnLayout.Step>

<TwoColumnLayout.Step title="Define the Drizzle schema">
<TwoColumnLayout.Block>

Create a TypeScript schema for a `posts` table. Drizzle uses this for both the migration and your type-safe queries.

</TwoColumnLayout.Block>
<TwoColumnLayout.Block>

```typescript filename="lib/db/schema.ts"
import { bigint, boolean, pgTable, text, timestamp } from 'drizzle-orm/pg-core';

export const posts = pgTable('posts', {
  id: bigint('id', { mode: 'number' })
    .primaryKey()
    .generatedByDefaultAsIdentity(),
  userId: text('user_id').notNull(),
  content: text('content').notNull(),
  isPublished: boolean('is_published').notNull().default(false),
  createdAt: timestamp('created_at', { withTimezone: true })
    .notNull()
    .defaultNow(),
});
```

```typescript filename="drizzle.config.ts"
import { loadEnvConfig } from '@next/env';
import { defineConfig } from 'drizzle-kit';

// drizzle-kit is a standalone CLI, so it doesn't pick up .env.local automatically.

> // loadEnvConfig matches Next.js's env loading behavior.

Reviewer: felt these comments unnecessary

```suggestion
```

loadEnvConfig(process.cwd());

export default defineConfig({
  schema: './lib/db/schema.ts',
  dialect: 'postgresql',
  dbCredentials: {
    url: process.env.DATABASE_URL!,
  },
});
```

<Admonition type="note">
`bigint('id', { mode: 'number' })` returns a JavaScript `number`. Postgres `bigint` can hold values above `Number.MAX_SAFE_INTEGER` (2^53 - 1), so for tables that may grow past about 9 quadrillion rows, switch to `{ mode: 'bigint' }` to keep precision.

> </Admonition>

Reviewer: felt this also unrelated for this guide

```suggestion
```

</TwoColumnLayout.Block>
</TwoColumnLayout.Step>

<TwoColumnLayout.Step title="Push the schema and seed sample data">
<TwoColumnLayout.Block>

> `drizzle-kit push` creates the table directly from your schema. For production you'd switch to generated migrations, but push is faster for a tutorial.

Reviewer: just a minor suggestion:
> `drizzle-kit push` creates the table directly from your schema. In production, you’d typically use `drizzle-kit generate` and `drizzle-kit migrate` instead.

    <main className="p-8">
      <h1 className="mb-4 text-2xl font-bold">Published posts</h1>
      <ul className="space-y-2">
        {allPosts.map((post) => (
          <li key={post.id} className="rounded border p-3">
            {post.content}
          </li>
        ))}
      </ul>
    </main>
  );
}
```

</TwoColumnLayout.Block>
</TwoColumnLayout.Step>

<TwoColumnLayout.Step title="Run the app">
<TwoColumnLayout.Block>

Start the dev server, then open [http://localhost:3000/auth/sign-up](http://localhost:3000/auth/sign-up). Create a test user, and you'll be redirected to `/posts` where the two seeded posts appear.

If you visit `/posts` without signing in, the middleware redirects you to `/auth/sign-in`.

</TwoColumnLayout.Block>
<TwoColumnLayout.Block>

```bash filename="Terminal"
npm run dev
```

</TwoColumnLayout.Block>
</TwoColumnLayout.Step>

</TwoColumnLayout>

## What you built

You now have a Next.js app where:

- Sign-up and sign-in are handled by Neon Auth via server actions that call `auth.signUp.email()` and `auth.signIn.email()`
- The `/posts` route is protected by middleware
- Published posts are queried server-side via Drizzle with full TypeScript types

> - Deploys to any Next.js App Router host that supports server actions, including Vercel, Netlify, and self-hosted Node

Reviewer: ```suggestion
- The application can be deployed to any Next.js App Router host that supports server actions, including Vercel, Netlify, and self-hosted Node
```

export async function signInWithEmail(
  _prev: { error: string } | null,
  formData: FormData,
) {
  const { error } = await auth.signIn.email({
    email: formData.get('email') as string,
    password: formData.get('password') as string,
  });

  if (error) return { error: error.message || 'Failed to sign in' };

  redirect('/posts');
}
```

</TwoColumnLayout.Block>
</TwoColumnLayout.Step>

<TwoColumnLayout.Step title="Query Postgres from a Server Component">
<TwoColumnLayout.Block>

Create the Drizzle client and a protected `/posts` page. The page is a Server Component, so both the session lookup and the Drizzle query run on the server at request time. `auth.getSession()` reads the signed-in user from cookies, Drizzle returns typed query results, and `dynamic = 'force-dynamic'` keeps the data fresh on every request.

</TwoColumnLayout.Block>
<TwoColumnLayout.Block>

```typescript filename="lib/db/client.ts"
import { drizzle } from 'drizzle-orm/neon-http';
import { neon } from '@neondatabase/serverless';
import * as schema from './schema';

const sql = neon(process.env.DATABASE_URL!);
export const db = drizzle(sql, { schema });
```

```tsx filename="app/posts/page.tsx"

> import { auth } from '@/lib/auth/server';
> import { db } from '@/lib/db/client';
> import { posts } from '@/lib/db/schema';
> import { desc, eq } from 'drizzle-orm';
> 
> export const dynamic = 'force-dynamic';
> 
> export default async function PostsPage() {
>   const { data: session } = await auth.getSession();
> 
>   const allPosts = await db
>     .select()
>     .from(posts)
>     .where(eq(posts.isPublished, true))
>     .orderBy(desc(posts.createdAt))
>     .limit(10);
> 
>   return (
>     <main className="p-8">
>       <h1 className="mb-1 text-2xl font-bold">Published posts</h1>
>       {session?.user && (
>         <p className="mb-4 text-sm text-gray-600">
>           Signed in as <span className="font-medium">{session.user.name}</span>
>         </p>
>       )}
>       <ul className="space-y-2">
>         {allPosts.map((post) => (
>           <li key={post.id} className="rounded border p-3">
>             {post.content}
>           </li>
>         ))}
>       </ul>
>     </main>
>   );
> }

Reviewer: ```suggestion
import { auth } from '@/lib/auth/server';
import { db } from '@/lib/db/client';
import { posts } from '@/lib/db/schema';
import { desc, eq } from 'drizzle-orm';

export const dynamic = 'force-dynamic';

export default async function PostsPage() {
  const { data: session } = await auth.getSession();

  const allPosts = await db
    .select()
    .from(posts)
    .where(eq(posts.isPublished, true))
    .orderBy(desc(posts.createdAt))
    .limit(10);

  return (
    <main className="p-8">
      <div className="mb-6">
        <h1 className="text-3xl font-bold tracking-tight text-gray-900 dark:text-white">
          Published posts
        </h1>

        {session?.user && (
          <p className="mt-2 text-sm text-gray-600 dark:text-gray-400">
            Signed in as{' '}
            <span className="font-medium text-gray-900 dark:text-gray-100">
              {session.user.name}
            </span>
          </p>
        )}
      </div>

      <ul className="space-y-3">
        {allPosts.map((post) => (
          <li
            key={post.id}
            className="rounded-xl border border-gray-200 bg-white p-4 shadow-sm transition-colors dark:border-gray-800 dark:bg-gray-900"
          >
            <p className="text-sm leading-6 text-gray-700 dark:text-gray-300">
              {post.content}
            </p>
          </li>
        ))}
      </ul>
    </main>
  );
}
```

felt theme would be good with this change (not important, can ignore)

<img width="1269" height="470" alt="Image" src="https://github.com/user-attachments/assets/905bc953-7b03-4353-b55b-57846fc52424" />

<img width="1266" height="419" alt="Image" src="https://github.com/user-attachments/assets/4c1dee79-ed7a-4c99-a3e1-672aa48b3faa" />
